File: threedollar-api/src/test/java/com/depromeet/threedollar/api/service/visit/VisitHistoryServiceTest.java
package com.depromeet.threedollar.api.service.visit;

import com.depromeet.threedollar.api.controller.medal.UserMedalEventListener;
import com.depromeet.threedollar.domain.event.visit.VisitHistoryAddedEvent;
import com.depromeet.threedollar.api.service.SetupStoreServiceTest;
import com.depromeet.threedollar.api.service.visit.dto.request.AddVisitHistoryRequest;
import com.depromeet.threedollar.common.exception.model.ConflictException;
import com.depromeet.threedollar.common.exception.model.NotFoundException;
import com.depromeet.threedollar.domain.domain.visit.VisitHistory;
import com.depromeet.threedollar.domain.domain.visit.VisitHistoryCreator;
import com.depromeet.threedollar.domain.domain.visit.VisitHistoryRepository;
import com.depromeet.threedollar.domain.domain.visit.VisitType;
import org.javaunit.autoparams.AutoSource;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;

import java.time.LocalDate;
import java.util.List;

import static com.depromeet.threedollar.api.assertutils.assertVisitHistoryUtils.assertVisitHistory;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@SpringBootTest
class VisitHistoryServiceTest extends SetupStoreServiceTest {

    @Autowired
    private VisitHistoryService visitHistoryService;

    @Autowired
    private VisitHistoryRepository visitHistoryRepository;

    @MockBean
    private UserMedalEventListener userMedalEventListener;

    @AfterEach
    void cleanUp() {
        visitHistoryRepository.deleteAllInBatch();
        super.cleanup();
    }

    @DisplayName("가게 방문 인증 등록")
    @Nested
    class AddStoreVisitHistory {

        @AutoSource
        @ParameterizedTest
        void 유저가_가게_방문_인증_정보를_추가한다(LocalDate dateOfVisit, VisitType visitType) {
            // given
            AddVisitHistoryRequest request = AddVisitHistoryRequest.testInstance(storeId, visitType);

            // when
            visitHistoryService.addVisitHistory(request, userId, dateOfVisit);

            // then
            List<VisitHistory> histories = visitHistoryRepository.findAll();
            assertAll(
                () -> assertThat(histories).hasSize(1),
                () -> assertVisitHistory(histories.get(0), storeId, userId, visitType, dateOfVisit)
            );
        }

        @Test
        void 존재하지_않는_가게에_방문_인증시_NotFound_에러가_발생한다() {
            // given
            Long notFoundStoreId = -1L;
            AddVisitHistoryRequest request = AddVisitHistoryRequest.testInstance(notFoundStoreId, VisitType.EXISTS);

            // when & then
            assertThatThrownBy(() -> visitHistoryService.addVisitHistory(request, userId, LocalDate.of(2021, 11, 18))).isInstanceOf(NotFoundException.class);
        }

        @AutoSource
        @ParameterizedTest
        void 가게_방문_인증시_오늘_이미_방문인증한_가게면_Conflict_에러가_발생한다(LocalDate dateOfVisit, VisitType visitType) {
            // given
            visitHistoryRepository.save(VisitHistoryCreator.create(store, userId, VisitType.EXISTS, dateOfVisit));

            AddVisitHistoryRequest request = AddVisitHistoryRequest.testInstance(storeId, visitType);

            // when & then
            assertThatThrownBy(() -> visitHistoryService.addVisitHistory(request, userId, dateOfVisit)).isInstanceOf(ConflictException.class);
        }

        @AutoSource
        @ParameterizedTest
        void 가게_방문시_메달을_획득하는_작업이_수행된다(LocalDate dateOfVisit, VisitType visitType) {
            // given
            AddVisitHistoryRequest request = AddVisitHistoryRequest.testInstance(storeId, visitType);

            // when
            visitHistoryService.addVisitHistory(request, userId, dateOfVisit);

            // then
            verify(userMedalEventListener, times(1)).addObtainableMedalsByVisitStore(any(VisitHistoryAddedEvent.class));
        }

    }

}

File: threedollar-api/src/test/java/com/depromeet/threedollar/api/service/visit/VisitHistoryServiceTest.kt
package com.depromeet.threedollar.api.service.visit

import com.depromeet.threedollar.api.service.StoreSetupTest
import com.depromeet.threedollar.api.service.visit.dto.request.AddVisitHistoryRequest
import com.depromeet.threedollar.common.exception.model.ConflictException
import com.depromeet.threedollar.common.exception.model.NotFoundException
import com.depromeet.threedollar.domain.domain.visit.VisitHistory
import com.depromeet.threedollar.domain.domain.visit.VisitHistoryCreator
import com.depromeet.threedollar.domain.domain.visit.VisitHistoryRepository
import com.depromeet.threedollar.domain.domain.visit.VisitType
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.*
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.TestConstructor
import java.time.LocalDate

@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
@SpringBootTest
internal class VisitHistoryServiceTest(
    private val visitHistoryService: VisitHistoryService,
    private val visitHistoryRepository: VisitHistoryRepository
) : StoreSetupTest() {

    @AfterEach
    fun cleanUp() {
        super.cleanup()
        visitHistoryRepository.deleteAll()
    }

    @DisplayName("가게 방문 인증 등록")
    @Nested
    inner class AddStoreVisitHistory {

        @Test
        fun `유저가 가게 방문 인증을 등록하면 DB에 방문 정보가 기록된다`() {
            // given
            val request = AddVisitHistoryRequest(storeId, VisitType.EXISTS_STORE)

            // when
            visitHistoryService.addVisitHistory(request, userId)

            // then
            val histories = visitHistoryRepository.findAll()
            assertAll({
                assertThat(histories).hasSize(1)
                assertVisitHistory(
                    visitHistory = histories[0],
                    storeId = storeId,
                    userId = userId,
                    type = request.type,
                    dateOfVisit = LocalDate.now()
                )
            })
        }

        @Test
        fun `방문 인증시 존재하지 않은 가게인 경우 NotFoundException이 발생한다`() {
            // given
            val notFoundStoreId = 999L
            val request = AddVisitHistoryRequest(notFoundStoreId, VisitType.EXISTS_STORE)

            // when & then
            assertThatThrownBy {
                visitHistoryService.addVisitHistory(
                    request,
                    userId
                )
            }.isInstanceOf(NotFoundException::class.java)
        }

        @Test
        fun `이미 해당 유저가 오늘 방문한 가게인 경우 ConflictException이 발생한다`() {
            // given
            val dateOfVisit = LocalDate.of(2021, 10, 18)
            visitHistoryRepository.save(
                VisitHistoryCreator.create(
                    storeId,
                    userId,
                    VisitType.EXISTS_STORE,
                    dateOfVisit
                )
            )

            // when & then
            assertThatThrownBy {
                visitHistoryService.addVisitHistory(
                    AddVisitHistoryRequest(storeId, VisitType.NOT_EXISTS_STORE),
                    userId
                )
            }.isInstanceOf(ConflictException::class.java)
        }
    }

    private fun assertVisitHistory(
        visitHistory: VisitHistory,
        storeId: Long,
        userId: Long,
        type: VisitType,
        dateOfVisit: LocalDate
    ) {
        assertThat(visitHistory.storeId).isEqualTo(storeId)
        assertThat(visitHistory.userId).isEqualTo(userId)
        assertThat(visitHistory.type).isEqualTo(type)
        assertThat(visitHistory.dateOfVisit).isEqualTo(dateOfVisit)
    }

}
